### PR TITLE
Endpoint state logs

### PR DESCRIPTION
**Pull request by raybejjani**
**Summary of changes**:
These changes split out the endpoint status log into the out put of `cilium endpoint log`, a new command. `cilium endpoint get` no longer includes this data.
We also now include each state transition, with a reason, in this log. These reasons are a bit too generic but generally present what's happening with the endpoint. The default format is intended to help with grep, and we keep the older json scheme with `--json`. It might be better to switch this mode to the same structured log format we have for the agent logs, as that might mesh well with any analysis tooling (or it could use the javascript).
We also log failed state transitions, which we use to avoid certain transitions. This might be alarming to a user, however, since they are labelled warning but are normal operation. We can choose to log them only during debug or on some other toggle (or just to the agent logs).

**Example Output**
```
$ cilium endpoint log 175
Timestamp              Status    State                   Message
2017-11-16T16:00:30Z   OK        ready                   Endpoint policy updated
2017-11-16T16:00:30Z   OK        ready                   Successfully regenerated endpoint program due to endpoint policy updated & changes were needed
2017-11-16T16:00:30Z   OK        ready                   Completed endpoint regeneration with no pending regeneration requests
2017-11-16T16:00:25Z   OK        regenerating            Regenerating Endpoint BPF: endpoint policy updated & changes were needed
2017-11-16T16:00:25Z   OK        waiting-to-regenerate   Triggering endpoint regeneration due to policy updates
2017-11-16T16:00:25Z   OK        ready                   Set identity for this endpoint
2017-11-16T16:00:25Z   OK        waiting-for-identity    Endpoint policy update skipped because no changes were needed
2017-11-16T16:00:25Z   OK        waiting-for-identity    Successfully regenerated endpoint program due to Create endpoint from API PUT
2017-11-16T16:00:25Z   Warning   waiting-for-identity    Skipped invalid state transition to ready due to: Completed endpoint regeneration with no pending regeneration requests
2017-11-16T16:00:23Z   Warning   waiting-for-identity    Skipped invalid state transition to waiting-to-regenerate due to: Create endpoint from API PUT
```
```
cilium endpoint get 175
[
  {
    "addressing": {
      "ipv4": "10.0.42.254",
      "ipv6": "f00d::a0b:0:0:af"
    },
    "container-id": "41b3fe49b5d33d5cfad5975632ff1938d1617832fbe6ee44ca65ac1986798315",
    "container-name": "k8s_POD_app1-bbbd54bb7-9q7bd_default_406c889f-cae7-11e7-bedb-08002771f4c0_0",
    "host-mac": "ea:57:7f:12:87:05",
    "id": 175,
    "identity": {
      "id": 273,
      "labels": [
        "k8s:id=app1",
        "k8s:io.kubernetes.pod.namespace=default"
      ]
    },
    "interface-index": 183,
    "interface-name": "lxc41b3f",
    "labels": {
      "custom": [],
      "disabled": [],
      "orchestration-identity": [
        "k8s:io.kubernetes.pod.namespace=default",
        "k8s:id=app1"
      ],
      "orchestration-info": [
        "container:io.kubernetes.pod.uid=406c889f-cae7-11e7-bedb-08002771f4c0",
        "container:annotation.kubernetes.io/config.seen=2017-11-16T08:00:22.609062381-08:00",
        "k8s:pod-template-hash=666810663",
        "container:io.kubernetes.container.name=POD",
        "container:annotation.kubernetes.io/config.source=api",
        "container:annotation.kubernetes.io/created-by={\"kind\":\"SerializedReference\",\"apiVersion\":\"v1\",\"reference\":{\"kind\":\"ReplicaSet\",\"namespace\":\"default\",\"name\":\"app1-bbbd54bb7\",\"uid\":\"405f98c1-cae7-11e7-bedb-08002771f4c0\",\"apiVersion\":\"extensions\",\"resourceVersion\":\"122594\"}}\n",
        "container:io.kubernetes.pod.name=app1-bbbd54bb7-9q7bd",
        "container:io.kubernetes.docker.type=podsandbox"
      ]
    },
    "mac": "4a:b4:a0:5b:d5:76",
    "pod-name": "default:app1-bbbd54bb7-9q7bd",
    "policy": {
      "allowed-consumers": [
        1
      ],
      "build": 2,
      "cidr-policy": {
        "egress": [],
        "ingress": []
      },
      "id": 273,
      "l4": {
        "egress": [],
        "ingress": []
      }
    },
    "policy-enabled": false,
    "policy-revision": 2,
    "state": "ready"
  }
]
```

Partially addresses #1952 
Fixes: #1277 

```release-note
Added cilium endpoint log command that returns the endpoint's status log
Change endpoint status log in cilium endpoint get to show only the most recent log
```

**Review Notes**
1. I've broken backwards compatibility by removing the "status"  field from `cilium endpoint get`. We can retain some compatibility by only returning the most recent entry as a 1-element array.
2. I've threaded a reason through a bunch of places and it's not very elegant. This will eventually morph into a context object we can use to track why something is happening back to a root cause/event (e.g. periodic regeneration, event from etcd, event from orchestration framework, API/user event). I'm not sure if it's better to have a 1-element context in this PR, or have something more robust later.
3. We log failed state transitions as a warning but they are part of normal operation. Perhaps we should not?
4. Can I regenerate the Docs for the API change in this PR?

**How to test (optional)**:
Tests pass.
`make clean && make generate-api && make reload`
Note: There might be new/old files left in the api directory. In that case:
`rm -rf ./api/ && git checkout -f`

### CODE REVIEW

File: pkg/endpoint/endpoint.go
Comment (aanm) [inline pkg/endpoint/endpoint.go]: Put this method private.

**Review by aanm — COMMENTED** [review]
Looks super good, just some small nit

File: pkg/endpoint/endpoint.go
Comment (raybejjani) [reply pkg/endpoint/endpoint.go]: I can't :( It's called twice by endpointmanager (and only once by endpoint!).

I could replace all the calls with `LogStatus(typ, OK, msg)` instead. What do you think?

File: pkg/endpoint/endpoint.go
Comment (joestringer) [inline pkg/endpoint/endpoint.go:1010]: I realise this is being refactored, but it'd be nice to standardize timestamp output to RFC3339 format - should be time.Now().UTC().Format(time.RFC3339)

**Review by joestringer — APPROVED** [review]
This looks really good, thanks! It bothered me a bit that status included this potentially giant neverending log so this is a welcome change.

From CLI perspective I'm happy with this. We might consider putting the "--json" flag parsing into `cilium/cmd/root.go` then just reusing the common flag for not only this tool, but also for other tools - but that could be a followup. I filed issue #2083 to track this.

Your review comments highlighted the main things I noticed in other areas, so I'll say LGTM with the caveat that I assume you'll follow up on those outstanding review questions.

File: pkg/endpoint/endpoint.go
Comment (tgraf) [reply pkg/endpoint/endpoint.go]: Let's keep it as-is. We already noticed that the separation of endpoint and endpointmanager is creating a lot of complication. It seems like the right step to move them to a single package. They depend on each other. Once that's done, issues such as this will resolve itself.

File: pkg/endpoint/endpoint.go
Comment (raybejjani) [reply pkg/endpoint/endpoint.go:1010]: I think this is already the case, but it happens when encoding into the swagger wire-format (https://github.com/cilium/cilium/blob/ecd3d3f8f05cd31d59b3ebece520a6ff9c69d904/pkg/endpoint/endpoint.go#L657) (this was already the case). Were you seeing somewhere it was printed differently?

**Review by raybejjani — COMMENTED** [review]
@joestringer I would love to have all the CLI consistent with the `--json` toggle! This would then allow folks to use the CLI to access the API for other programmatic things. Issue #2083 is surprisingly large though :/
For the other open questions:
1. I've added `status` back to endpoint get but it only shows the most recent log. This won't break anything parsing the code, and should serve as simple indicator. I can remove it if you think it's a bad idea.
2. I think I'll stick with the string being passed around as it's least involved change. I will be adding event UUID tracking and that change will replace this reason field with something more relevant.
3. I'm still logging these since there isn't a straight forward way to hide them without losing that data. We also log them to the agent logs as info, so they will already be seen.
4. I realised I already knew how to do this, and that nothing needs to be in this PR.


@tgraf Assuming no one raises an object, this can be merged (and I'll add the tag once an API owner approves).

File: daemon/endpoint.go
Comment (tgraf) [inline daemon/endpoint.go:268]: I think several of these reasons should become more end user friendly going forward but I understand the desire to provide specific context and reasoning so I'm fine to start with this and eventually add a more end user focused reason summary later.

**Review by tgraf — APPROVED** [review]
1. The workaround to provide compatibility with scripts is perfect 👍 

2. I'm totally fine with the current approach to passing around reason as string. I think we should discuss and decide to add `context` universally before we just bluntly walk into it.

3. I'm fine with the logs as-is. There is room for improvement here but can be done outside of this PR.

4. Good to see that you figured out that the API reference docs are automatically generated from the swappger spec.

File: pkg/endpoint/endpoint.go
Comment (joestringer) [reply pkg/endpoint/endpoint.go:1010]: Ah, good. I wasn't observing any different, just that some of the usages seem to explicitly format to RFC3339, while some don't (just using 'grep' across the tree for similar usages).

**Pull request merged**
